File: packages/python/src/omniharness/providers/_devin_v3.py
from __future__ import annotations

import logging
from datetime import datetime, timezone
from typing import Any, AsyncIterator, Optional

import httpx

from .._streaming import polling_stream
from ..errors import ProviderError, classify_http_error
from ..types import (
    Artifact,
    Event,
    ProviderId,
    ProviderCapabilities,
    ReplayToken,
    Session,
    SessionOptions,
    SessionPage,
    SessionStatus,
)

_log = logging.getLogger(__name__)

_STATUS_MAP: dict[str, SessionStatus] = {
    "running": "in_progress",
    "exit": "completed",
    "error": "failed",
    "suspended": "paused",
}


def _parse_dt(s: str) -> datetime:
    return datetime.fromisoformat(s.replace("Z", "+00:00"))


class DevinV3Provider:
    id: ProviderId = "devin"
    capabilities = ProviderCapabilities(
        streaming=False,
        cancellation=True,
        list_sessions=True,
        send_message=True,
        plan_approval=False,
        tool_results=False,
        file_upload=False,
    )

    def __init__(self, api_key: str, org_id: str) -> None:
        if not org_id:
            raise ProviderError(
                "invalid_request",
                "DEVIN_ORG_ID is required for the Devin provider",
                "devin",
                False,
            )
        self._org_id = org_id
        self._http = httpx.AsyncClient(
            base_url=f"https://api.devin.ai/v3/organizations/{org_id}",
            headers={"Authorization": f"Bearer {api_key}"},
            timeout=30.0,
        )

    async def _req(self, method: str, path: str, **kwargs: Any) -> httpx.Response:
        resp = await self._http.request(method, path, **kwargs)
        if resp.status_code >= 400:
            raise classify_http_error(
                resp.status_code,
                resp.json() if resp.content else {},
                "devin",
                dict(resp.headers),
            )
        return resp

    def _build_prompt(self, opts: SessionOptions) -> str:
        po = opts.provider_options or {}
        parts: list[str] = []
        if opts.source and opts.source.uri != "none:":
            parts.append(f"Repository: {opts.source.uri}\n")
        skills = po.get("skills", [])
        if skills:
            skill_blocks = ["---", "# Context"]
            for sk in skills:
                skill_blocks.append(f"## {sk['name']}\n{sk['content']}")
            skill_blocks.append("---")
            parts.append("\n".join(skill_blocks) + "\n")
        parts.append(opts.prompt)
        if po.get("auto_create_pr"):
            parts.append(
                "\n\nWhen complete, open a pull request with your changes."
            )
        return "".join(parts)

    def _to_session(self, data: dict[str, Any]) -> Session:
        session_id = data.get("session_id", data.get("id", ""))
        raw_status = data.get("status", "running")
        status: SessionStatus = _STATUS_MAP.get(raw_status, "in_progress")
        artifacts: list[Artifact] = []
        for att in data.get("attachments", []):
            artifacts.append(
                Artifact(
                    kind="file",
                    provider="devin",
                    url=att.get("url"),
                    name=att.get("name"),
                    data=att,
                )
            )
        if pr_url := data.get("pull_request_url"):
            artifacts.append(
                Artifact(kind="pull_request", provider="devin", url=pr_url)
            )
        token = ReplayToken._encode(
            {
                "session_id": session_id,
                "org_id": self._org_id,
                "url": data.get("url", ""),
            }
        )
        ts_str = data.get("created_at", datetime.now(timezone.utc).isoformat())
        return Session(
            id=session_id,
            provider="devin",
            status=status,
            created_at=_parse_dt(ts_str),
            updated_at=_parse_dt(data.get("updated_at", ts_str)),
            artifacts=artifacts,
            replay_token=token,
        )

    async def create_session(self, opts: SessionOptions) -> Session:
        po = opts.provider_options or {}
        if po.get("mcp_servers"):
            _log.debug("devin: mcp_servers in provider_options is not supported and will be ignored")
        body: dict[str, Any] = {"prompt": self._build_prompt(opts)}
        if opts.idempotency_key:
            body["idempotency_key"] = opts.idempotency_key
        if create_as := po.get("create_as_user_id"):
            body["create_as_user_id"] = create_as
        resp = await self._req("POST", "/sessions", json=body)
        return self._to_session(resp.json())

    async def get_session(
        self,
        session_id: str,
        replay_token: Optional[ReplayToken] = None,
    ) -> Session:
        resp = await self._req("GET", f"/sessions/{session_id}")
        return self._to_session(resp.json())

    async def stream_events(
        self,
        session_id: str,
        replay_token: Optional[ReplayToken] = None,
        since_seq: int = 0,
    ) -> AsyncIterator[Event]:
        seen_msg_indices: set[int] = set()
        seen_attachment_ids: set[str] = set()

        async def fetch_session() -> Session:
            return await self.get_session(session_id, replay_token)

        async def fetch_new_events(
            cursor: Optional[str],
        ) -> tuple[list[Event], Optional[str]]:
            events: list[Event] = []

            # Poll messages
            msg_resp = await self._req("GET", f"/sessions/{session_id}/messages")
            for i, msg in enumerate(msg_resp.json().get("messages", [])):
                if i in seen_msg_indices:
                    continue
                seen_msg_indices.add(i)
                text = msg.get("message", msg.get("text", ""))
                if text:
                    events.append(
                        Event(
                            kind="message",
                            session_id=session_id,
                            provider="devin",
                            ts=_parse_dt(
                                msg.get("created_at", datetime.now(timezone.utc).isoformat())
                            ),
                            seq=0,
                            data={"text": text},
                            raw=msg,
                        )
                    )

            # Poll attachments
            att_resp = await self._req("GET", f"/sessions/{session_id}/attachments")
            for att in att_resp.json().get("attachments", []):
                att_id = att.get("id", "")
                if att_id in seen_attachment_ids:
                    continue
                seen_attachment_ids.add(att_id)
                artifact = Artifact(
                    kind="file",
                    provider="devin",
                    url=att.get("url"),
                    name=att.get("name"),
                    data=att,
                )
                events.append(
                    Event(
                        kind="artifact",
                        session_id=session_id,
                        provider="devin",
                        ts=_parse_dt(
                            att.get("created_at", datetime.now(timezone.utc).isoformat())
                        ),
                        seq=0,
                        data={"artifact": artifact},
                        raw=att,
                    )
                )

            return events, None

        async for ev in polling_stream(
            provider="devin",
            session_id=session_id,
            fetch_session=fetch_session,
            fetch_new_events=fetch_new_events,
        ):
            yield ev

    async def cancel_session(
        self,
        session_id: str,
        replay_token: Optional[ReplayToken] = None,
    ) -> None:
        await self._req("DELETE", f"/sessions/{session_id}")

    async def send_message(
        self,
        session_id: str,
        message: str,
        replay_token: Optional[ReplayToken] = None,
    ) -> None:
        await self._req(
            "POST",
            f"/sessions/{session_id}/messages",
            json={"message": message},
        )

    async def list_sessions(
        self,
        limit: int = 20,
        cursor: Optional[str] = None,
    ) -> SessionPage:
        params: dict[str, Any] = {"first": limit}
        if cursor:
            params["after"] = cursor
        resp = await self._req("GET", "/sessions", params=params)
        data = resp.json()
        sessions = [self._to_session(s) for s in data.get("sessions", [])]
        return SessionPage(
            sessions=sessions,
            next_cursor=data.get("end_cursor"),
            provider="devin",
        )

File: packages/typescript/src/providers/devin-v3.ts
import { classifyHttpError, ProviderError } from "../errors.js";
import { pollingStream } from "../streaming.js";
import type {
  Artifact,
  Event,
  ProviderId,
  ProviderCapabilities,
  ReplayToken,
  Session,
  SessionOptions,
  SessionPage,
  SessionStatus,
} from "../types.js";
import { ReplayToken as RT } from "../types.js";

const STATUS_MAP: Record<string, SessionStatus> = {
  running: "in_progress",
  exit: "completed",
  error: "failed",
  suspended: "paused",
};

async function req(
  apiKey: string,
  method: string,
  url: string,
  body?: unknown,
): Promise<Record<string, unknown>> {
  const resp = await fetch(url, {
    method,
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: body !== undefined ? JSON.stringify(body) : undefined,
  });
  const data = resp.ok ? ((await resp.json()) as Record<string, unknown>) : {};
  if (!resp.ok) {
    const hdrs: Record<string, string> = {};
    resp.headers.forEach((v, k) => { hdrs[k] = v; });
    throw classifyHttpError(resp.status, data, "devin", hdrs);
  }
  return data;
}

export class DevinV3Provider {
  readonly id: ProviderId = "devin";
  readonly capabilities: ProviderCapabilities = {
    streaming: false,
    cancellation: true,
    listSessions: true,
    sendMessage: true,
    planApproval: false,
    toolResults: false,
    fileUpload: false,
  };

  private readonly base: string;

  constructor(
    private readonly apiKey: string,
    readonly orgId: string,
  ) {
    if (!orgId) {
      throw new ProviderError(
        "invalid_request",
        "DEVIN_ORG_ID is required for the Devin provider",
        "devin",
        false,
      );
    }
    this.base = `https://api.devin.ai/v3/organizations/${orgId}`;
  }

  private buildPrompt(opts: SessionOptions): string {
    const po = opts.providerOptions ?? {};
    const parts: string[] = [];
    if (opts.source && opts.source.uri !== "none:") {
      parts.push(`Repository: ${opts.source.uri}\n`);
    }
    const skills = (po["skills"] as Array<{ name: string; content: string }> | undefined) ?? [];
    if (skills.length > 0) {
      const blocks = ["---", "# Context"];
      for (const sk of skills) {
        blocks.push(`## ${sk.name}\n${sk.content}`);
      }
      blocks.push("---");
      parts.push(blocks.join("\n") + "\n");
    }
    parts.push(opts.prompt);
    if (po["auto_create_pr"]) {
      parts.push("\n\nWhen complete, open a pull request with your changes.");
    }
    return parts.join("");
  }

  private toSession(data: Record<string, unknown>): Session {
    const sessionId =
      (data["session_id"] as string | undefined) ?? (data["id"] as string | undefined) ?? "";
    const rawStatus = (data["status"] as string | undefined) ?? "running";
    const status: SessionStatus = STATUS_MAP[rawStatus] ?? "in_progress";
    const artifacts: Artifact[] = [];
    for (const att of (data["attachments"] as Array<Record<string, unknown>> | undefined) ?? []) {
      artifacts.push({
        kind: "file",
        provider: "devin",
        url: att["url"] as string | undefined,
        name: att["name"] as string | undefined,
        data: att,
      });
    }
    const prUrl = data["pull_request_url"] as string | undefined;
    if (prUrl) artifacts.push({ kind: "pull_request", provider: "devin", url: prUrl });
    const token = RT._encode({
      session_id: sessionId,
      org_id: this.orgId,
      url: (data["url"] as string | undefined) ?? "",
    });
    const tsStr = (data["created_at"] as string | undefined) ?? new Date().toISOString();
    return {
      id: sessionId,
      provider: "devin",
      status,
      createdAt: new Date(tsStr).toISOString(),
      updatedAt: new Date((data["updated_at"] as string | undefined) ?? tsStr).toISOString(),
      artifacts,
      replayToken: token,
    };
  }

  async createSession(opts: SessionOptions): Promise<Session> {
    const po = opts.providerOptions ?? {};
    if (po["mcp_servers"]) {
      console.debug("devin: mcp_servers in providerOptions is not supported and will be ignored");
    }
    const body: Record<string, unknown> = { prompt: this.buildPrompt(opts) };
    if (opts.idempotencyKey) body["idempotency_key"] = opts.idempotencyKey;
    if (po["create_as_user_id"]) body["create_as_user_id"] = po["create_as_user_id"];
    const data = await req(this.apiKey, "POST", `${this.base}/sessions`, body);
    return this.toSession(data);
  }

  async getSession(sessionId: string, _replayToken?: ReplayToken): Promise<Session> {
    const data = await req(this.apiKey, "GET", `${this.base}/sessions/${sessionId}`);
    return this.toSession(data);
  }

  async *streamEvents(
    sessionId: string,
    replayToken?: ReplayToken,
    _sinceSeq = 0,
  ): AsyncIterableIterator<Event> {
    const seenMsgIndices = new Set<number>();
    const seenAttIds = new Set<string>();
    const self = this;

    const fetchSession = () => self.getSession(sessionId, replayToken);

    const fetchNewEvents = async (
      _cursor: string | undefined,
    ): Promise<[Event[], string | undefined]> => {
      const events: Event[] = [];

      const msgData = await req(self.apiKey, "GET", `${self.base}/sessions/${sessionId}/messages`);
      const messages = (msgData["messages"] as Array<Record<string, unknown>> | undefined) ?? [];
      for (let i = 0; i < messages.length; i++) {
        if (seenMsgIndices.has(i)) continue;
        seenMsgIndices.add(i);
        const msg = messages[i]!;
        const text = (msg["message"] as string | undefined) ?? (msg["text"] as string | undefined) ?? "";
        if (text) {
          events.push({
            kind: "message",
            sessionId,
            provider: "devin",
            ts: new Date((msg["created_at"] as string | undefined) ?? Date.now()).toISOString(),
            seq: 0,
            data: { text },
            raw: msg,
          });
        }
      }

      const attData = await req(self.apiKey, "GET", `${self.base}/sessions/${sessionId}/attachments`);
      const attachments = (attData["attachments"] as Array<Record<string, unknown>> | undefined) ?? [];
      for (const att of attachments) {
        const attId = (att["id"] as string | undefined) ?? "";
        if (seenAttIds.has(attId)) continue;
        seenAttIds.add(attId);
        const artifact: Artifact = {
          kind: "file",
          provider: "devin",
          url: att["url"] as string | undefined,
          name: att["name"] as string | undefined,
          data: att,
        };
        events.push({
          kind: "artifact",
          sessionId,
          provider: "devin",
          ts: new Date((att["created_at"] as string | undefined) ?? Date.now()).toISOString(),
          seq: 0,
          data: { artifact },
          raw: att,
        });
      }

      return [events, undefined];
    };

    yield* pollingStream({ provider: "devin", sessionId, fetchSession, fetchNewEvents });
  }

  async cancelSession(sessionId: string, _replayToken?: ReplayToken): Promise<void> {
    await req(this.apiKey, "DELETE", `${this.base}/sessions/${sessionId}`);
  }

  async sendMessage(sessionId: string, message: string, _replayToken?: ReplayToken): Promise<void> {
    await req(this.apiKey, "POST", `${this.base}/sessions/${sessionId}/messages`, { message });
  }

  async listSessions(limit = 20, cursor?: string): Promise<SessionPage> {
    const url = new URL(`${this.base}/sessions`);
    url.searchParams.set("first", String(limit));
    if (cursor) url.searchParams.set("after", cursor);
    const data = await req(this.apiKey, "GET", url.toString());
    const sessions = ((data["sessions"] as Array<Record<string, unknown>> | undefined) ?? []).map(
      (s) => this.toSession(s),
    );
    return { sessions, nextCursor: data["end_cursor"] as string | undefined, provider: "devin" };
  }
}
